assets/js-text-with-image.js

```diff
@@ -1,170 +1,222 @@
+  selector: {
+    section: '.text-image',
+    spacer: '.text-image__spacer'
+  },
+  palettes: ['palette-one', 'palette-two', 'palette-three']
+}
 
-  // Cache elements but get previous section on demand
-  const getElements = () => ({
-    section,
-    prevSection: $.getSibling(section, '*', 'prev')
-  })
+const TextImageDOM = {
+  elements: {
+    sections: null
+  },
 
-  // Cache checking functions for better performance
-  const hasPalette = (el, palette) => el.classList.contains(palette);
-  const hasSpacerClass = (el) => el.classList.contains(config.class.spacer);
+  cache: {
+    spacers: new Map(),
+    prevSections: new Map()
+  },
 
-  const applyPaletteToSpacer = (spacer, prevChild) => {
-    let foundPrevPalette = null,
-        currentSpacerPalette = null;
+  init() {
+    this.elements.sections = document.querySelectorAll(TextImageConfig.selector.section);
+    return this.elements.sections && this.elements.sections.length > 0;
+  },
 
-    const findPalette = (el) => {
-      for (let i = 0; i < config.class.palettes.length; i++) {
-        const palette = config.class.palettes[i];
+  getPreviousSection(section) {
+    if (!section) return null;
 
-        if (hasPalette(el, palette)) return palette;
-      }
-      return null;
+    if (this.cache.prevSections.has(section)) {
+      return this.cache.prevSections.get(section);
     }
 
-    foundPrevPalette = findPalette(prevChild); // Find palettes in previous child and current spacer
-
-    if (!foundPrevPalette) return;
-
-    currentSpacerPalette = findPalette(spacer); // Find the current palette in the spacer
-
-    // Apply palette class efficiently using Utils with frame sequencing
-    const applyClasses = () => {
-      if (currentSpacerPalette && currentSpacerPalette !== foundPrevPalette) {
-        // Use frameSequence for better performance and transition handling
-        $.frameSequence(
-          // Read phase - no actual reads needed here, just prepare data
-          () => ({
-            remove: currentSpacerPalette,
-            add: foundPrevPalette
-          }),
-          // Write phase - perform DOM updates with proper sequencing
-          (data) => {
-            $.toggleClass(spacer, data.remove, false);
-            $.nextFrame(() => $.toggleClass(spacer, data.add, true));
-          }
-        )
-      } else if (!currentSpacerPalette) {
-        // No palette on the spacer, just add the one from the previous section
-        $.nextFrame(() => $.toggleClass(spacer, foundPrevPalette, true));
-      }
+    const prevSection = $.getSibling(section, '*', 'prev');
+    this.cache.prevSections.set(section, prevSection);
+
+    return prevSection;
+  },
+
+  getSpacer(section) {
+    if (!section) return null;
+    if (this.cache.spacers.has(section)) {
+      return this.cache.spacers.get(section);
     }
 
-    $.batchDOM(applyClasses);
+    const spacer = section.querySelector('.' + TextImageConfig.selector.spacer.substring(1));
```

Comment:
We could use template strings here? 

```suggestion
    const spacer = section.querySelector(`.${TextImageConfig.selector.spacer.substring(1)`);
```